HousingRepairsOnlineApi/UseCases/BookAvailableAppointmentUseCase.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Ardalis.GuardClauses;
+using HousingRepairsOnlineApi.Gateways;
+using HousingRepairsOnlineApi.Helpers;
+
+namespace HousingRepairsOnlineApi.UseCases
+{
+    public class BookAvailableAppointmentUseCase : IBookAvailableAppointmentUseCase
+    {
+        private readonly IAppointmentsGateway appointmentsGateway;
+        private readonly IAppointmentSlotsFilter appointmentSlotsFilter;
+
+        public BookAvailableAppointmentUseCase(IAppointmentsGateway appointmentsGateway, IAppointmentSlotsFilter appointmentSlotsFilter)
+        {
+            this.appointmentsGateway = appointmentsGateway;
+            this.appointmentSlotsFilter = appointmentSlotsFilter;
+        }
+
+        public async Task Execute(string bookingReference, string sorCode, string priority, string locationId, string repairDescriptionText)
+        {
+            Guard.Against.NullOrWhiteSpace(bookingReference, nameof(sorCode));
+            Guard.Against.NullOrWhiteSpace(sorCode, nameof(sorCode));
+            Guard.Against.NullOrWhiteSpace(priority, nameof(priority));
+            Guard.Against.NullOrWhiteSpace(locationId, nameof(locationId));
+            Guard.Against.NullOrWhiteSpace(repairDescriptionText, nameof(repairDescriptionText));
+
+            var allowedAppointmentSlotTimeSpans = appointmentSlotsFilter.Filter();
+
+            var appointments = await appointmentsGateway.GetAvailableAppointments(sorCode, priority, locationId, null, allowedAppointmentSlotTimeSpans);
+            // Need to get next available appointment
+            var appointment = appointments.FirstOrDefault();
```

Comment:
`FirstOrDefault` would return null if `appointments` is empty (i.e. lack of availability).
We should expect that could happen and need to handle it gracefully.

Reply:
I have raised an exception where no appointments returned and tested.